torch/csrc/autograd/input_metadata.cpp

```diff
@@ -53,17 +53,67 @@ at::Tensor InputMetadata::zeros_like() const {
   return at::zeros_symint(shape_as_dim_vector(), options_);
 }
 
+static bool definitely_true(const c10::SymBool& b) {
+  return b.has_hint() && b.guard_bool(__FILE__, __LINE__);
+}
+
 at::Tensor InputMetadata::maybe_reduce(
     const size_t i,
     at::Tensor grad,
     const std::function<std::string(const std::string&)>& format_error) const {
-  if (!is_same_shape(grad)) {
-    if (is_expandable_to_shape(grad)) {
-      return reduce_grad(grad);
+  auto fail = [&]() {
+    const auto message = incompatible_shape_error_message(i, grad);
+    TORCH_CHECK(false, format_error(message.str()));
+  };
+
+  // Nested tensor makes my brain explode, so I've just hard-coded the logic
+  // for this case, at risk of code duplication.  This logic does NOT do the
+  // careful oblivious logic as seen below
+  if (is_nested_ || is_cpp_nested_tensor() || grad.is_nested() ||
+      ::torch::autograd::is_cpp_nested_tensor(grad)) {
+    if (!is_same_shape(grad)) {
+      if (is_expandable_to_shape(grad)) {
+        return reduce_grad(grad);
+      } else {
+        fail();
+      }
     } else {
-      const auto message = incompatible_shape_error_message(i, grad);
-      TORCH_CHECK(false, format_error(message.str()));
+      return grad;
     }
+  }
+
+  auto shape = shape_as_dim_vector();
+  auto desired = grad.sym_sizes();
+
+  size_t ndim = shape.size();
+  size_t target_dim = desired.size();
+  if (ndim > target_dim) {
+    fail();
+  }
+  bool needs_reduce = false;
+  for (const auto i : c10::irange(ndim)) {
```

Comment:
The problem is that `is_expandable_to` is returning a true/false quantity, which may or may not cause an error condition. However, I can only produce a deferred runtime assert if, at the time I try to guard on a boolean expression, I *know* that if this expression evaluates to false it will be an error condition. Which is_expandable_to doesn't! This is why everything needed to be inlined here.

Reply:
Is that what SymBool can solve?
But I guess I would need to understand the unbaked symint strategy to see what is feasible here.

Reply:
Returning a SymBool could help. However, now you have to rewrite is_expandable_to to be data oblivious: it's not allowed to do an early return when it notices that something is not expandable to.